migrations.js

```diff
@@ -3,33 +3,75 @@ var fs = require('fs'),
     async = require('async'),
     db = require('./db');
 
-var hasRun = function(meta, migration) {
-  if (!meta || !meta.migrations) {
+var MIGRATION_LOG_ID = 'migration-log',
+    MIGRATION_LOG_TYPE = 'meta';
+
+var hasRun = function(log, migration) {
+  if (!log || !log.migrations) {
     return false;
   }
-  return meta.migrations.indexOf(migration.name) !== -1;
+  return log.migrations.indexOf(migration.name) !== -1;
 };
 
 var error = function(migration, err) {
   return 'Migration "' + migration.name + '" failed with: ' + JSON.stringify(err);
 };
 
-var getMeta = function(callback) {
-  db.medic.view('medic-client', 'doc_by_type', { include_docs: true, key: [ 'meta' ] }, function(err, meta) {
+var getLogWithView = function(callback) {
+  var options = { include_docs: true, key: [ MIGRATION_LOG_TYPE ] };
+  db.medic.view('medic-client', 'doc_by_type', options, function(err, result) {
     if (err) {
       return callback(
         new Error(
           'Could not run migrations without doc_by_type view. Update ddoc. ' +
           JSON.stringify(err)));
     }
-    meta = meta && meta.rows && meta.rows[0] && meta.rows[0].doc;
-    if (!meta) {
-      meta = { type: 'meta' };
+    var log = result && result.rows && result.rows[0] && result.rows[0].doc;
+    callback(null, log);
+  });
+};
+
+var deleteOldLog = function(oldLog, callback) {
+  if (!oldLog) {
+    return callback();
+  }
+  oldLog._deleted = true;
+  db.medic.insert(oldLog, callback);
+};
+
+var createLog = function(callback) {
+  getLogWithView(function(err, oldLog) {
+    if (err) {
+      return callback(err);
+    }
+    var newLog = {
+      _id: MIGRATION_LOG_ID,
+      type: MIGRATION_LOG_TYPE,
+      migrations: (oldLog && oldLog.migrations) || []
+    };
+    db.medic.insert(newLog, function(err) {
+      if (err) {
+        return callback(err);
+      }
+      deleteOldLog(oldLog, function(err) {
+        if (err) {
+          return callback(err);
+        }
+        getLog(callback);
+      });
+    });
+  });
+};
+
+var getLog = function(callback) {
+  db.medic.get(MIGRATION_LOG_ID, function(err, doc) {
+    if (!err) {
+      return callback(null, doc);
     }
-    if (!meta.migrations) {
-      meta.migrations = [];
+    if (err.statusCode === 404) {
+      return createLog(callback);
```

Comment:
There are no promises in this code anywhere, mostly because we still use nano which is callback based. One day...

I've cleaned up the code a little to hopefully make it easier to read.